// @parcel/transformer-css/lib/CSSTransformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _sourceMap() {
  const data = _interopRequireDefault(require("@parcel/source-map"));

  _sourceMap = function () {
    return data;
  };

  return data;
}

function _plugin() {
  const data = require("@parcel/plugin");

  _plugin = function () {
    return data;
  };

  return data;
}

function _css() {
  const data = require("@parcel/css");

  _css = function () {
    return data;
  };

  return data;
}

function _utils() {
  const data = require("@parcel/utils");

  _utils = function () {
    return data;
  };

  return data;
}

function _browserslist() {
  const data = _interopRequireDefault(require("browserslist"));

  _browserslist = function () {
    return data;
  };

  return data;
}

function _nullthrows() {
  const data = _interopRequireDefault(require("nullthrows"));

  _nullthrows = function () {
    return data;
  };

  return data;
}

function _diagnostic() {
  const data = _interopRequireWildcard(require("@parcel/diagnostic"));

  _diagnostic = function () {
    return data;
  };

  return data;
}

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = new (_plugin().Transformer)({
  async loadConfig({
    config,
    options
  }) {
    let conf = await config.getConfigFrom(options.projectRoot + '/index', [], {
      packageKey: '@parcel/transformer-css'
    });
    return conf === null || conf === void 0 ? void 0 : conf.contents;
  },

  async transform({
    asset,
    config,
    options
  }) {
    // Normalize the asset's environment so that properties that only affect JS don't cause CSS to be duplicated.
    // For example, with ESModule and CommonJS targets, only a single shared CSS bundle should be produced.
    let env = asset.env;
    asset.setEnvironment({
      context: 'browser',
      engines: {
        browsers: asset.env.engines.browsers
      },
      shouldOptimize: asset.env.shouldOptimize,
      shouldScopeHoist: asset.env.shouldScopeHoist,
      sourceMap: asset.env.sourceMap
    });
    let [code, originalMap] = await Promise.all([asset.getBuffer(), asset.getMap()]);
    let targets = getTargets(asset.env.engines.browsers);
    let res;

    try {
      if (asset.meta.type === 'attr') {
        res = (0, _css().transformStyleAttribute)({
          code,
          analyzeDependencies: true,
          targets
        });
      } else {
        let cssModules = false;

        if (asset.meta.type !== 'tag' && asset.meta.cssModulesCompiled == null) {
          let cssModulesConfig = config === null || config === void 0 ? void 0 : config.cssModules;

          if (asset.isSource && (typeof cssModulesConfig === 'boolean' || cssModulesConfig !== null && cssModulesConfig !== void 0 && cssModulesConfig.global) || /\.module\./.test(asset.filePath)) {
            if (cssModulesConfig !== null && cssModulesConfig !== void 0 && cssModulesConfig.dashedIdents && !asset.isSource) {
              cssModulesConfig.dashedIdents = false;
            }

            cssModules = cssModulesConfig !== null && cssModulesConfig !== void 0 ? cssModulesConfig : true;
          }
        }

        res = (0, _css().transform)({
          filename: _path().default.relative(options.projectRoot, asset.filePath),
          code,
          cssModules,
          analyzeDependencies: asset.meta.hasDependencies !== false,
          sourceMap: !!asset.env.sourceMap,
          drafts: config === null || config === void 0 ? void 0 : config.drafts,
          pseudoClasses: config === null || config === void 0 ? void 0 : config.pseudoClasses,
          targets
        });
      }
    } catch (err) {
      var _err$data;

      err.filePath = asset.filePath;
      let diagnostic = (0, _diagnostic().errorToDiagnostic)(err, {
        origin: '@parcel/transformer-css'
      });

      if (((_err$data = err.data) === null || _err$data === void 0 ? void 0 : _err$data.type) === 'AmbiguousUrlInCustomProperty' && err.data.url) {
        let p = '/' + (0, _utils().relativePath)(options.projectRoot, _path().default.resolve(_path().default.dirname(asset.filePath), err.data.url), false);
        diagnostic[0].hints = [`Replace with: url(${p})`];
        diagnostic[0].documentationURL = 'https://parceljs.org/languages/css/#url()';
      }

      throw new (_diagnostic().default)({
        diagnostic
      });
    }

    asset.setBuffer(res.code);

    if (res.map != null) {
      let vlqMap = JSON.parse(res.map.toString());
      let map = new (_sourceMap().default)(options.projectRoot);
      map.addVLQMap(vlqMap);

      if (originalMap) {
        map.extends(originalMap);
      }

      asset.setMap(map);
    }

    if (res.dependencies) {
      for (let dep of res.dependencies) {
        let loc = dep.loc;

        if (originalMap) {
          loc = (0, _utils().remapSourceLocation)(loc, originalMap);
        }

        if (dep.type === 'import' && !res.exports) {
          asset.addDependency({
            specifier: dep.url,
            specifierType: 'url',
            loc,
            meta: {
              // For the glob resolver to distinguish between `@import` and other URL dependencies.
              isCSSImport: true,
              media: dep.media
            }
          });
        } else if (dep.type === 'url') {
          asset.addURLDependency(dep.url, {
            loc,
            meta: {
              placeholder: dep.placeholder
            }
          });
        }
      }
    }

    let assets = [asset];

    if (res.exports != null) {
      let exports = res.exports;
      asset.symbols.ensure();
      asset.symbols.set('default', 'default');
      let dependencies = new Map();
      let locals = new Map();
      let c = 0;
      let depjs = '';
      let js = '';

      for (let key in exports) {
        locals.set(exports[key].name, key);
      }

      let seen = new Set();

      let add = key => {
        if (seen.has(key)) {
          return;
        }

        seen.add(key);
        let e = exports[key];
        let s = `module.exports[${JSON.stringify(key)}] = \`${e.name}`;

        for (let ref of e.composes) {
          s += ' ';

          if (ref.type === 'local') {
            add((0, _nullthrows().default)(locals.get(ref.name)));
            s += '${' + `module.exports[${JSON.stringify((0, _nullthrows().default)(locals.get(ref.name)))}]` + '}';
          } else if (ref.type === 'global') {
            s += ref.name;
          } else if (ref.type === 'dependency') {
            let d = dependencies.get(ref.specifier);

            if (d == null) {
              d = `dep_${c++}`;
              depjs += `import * as ${d} from ${JSON.stringify(ref.specifier)};\n`;
              dependencies.set(ref.specifier, d);
            }

            s += '${' + `${d}[${JSON.stringify(ref.name)}]` + '}';
          }
        }

        s += '`;\n'; // If the export is referenced internally (e.g. used @keyframes), add a self-reference
        // to the JS so the symbol is retained during tree-shaking.

        if (e.isReferenced) {
          s += `module.exports[${JSON.stringify(key)}];\n`;
        }

        js += s;
      };

      for (let key in exports) {
        asset.symbols.set(key, exports[key].name);
        add(key);
      }

      if (res.dependencies) {
        for (let dep of res.dependencies) {
          if (dep.type === 'import') {
            // TODO: Figure out how to treeshake this
            let d = `dep_$${c++}`;
            depjs += `import * as ${d} from ${JSON.stringify(dep.url)};\n`;
            js += `for (let key in ${d}) { if (key in module.exports) module.exports[key] += ' ' + ${d}[key]; else module.exports[key] = ${d}[key]; }\n`;
            asset.symbols.set('*', '*');
          }
        }
      }

      if (res.references != null) {
        let references = res.references;

        for (let symbol in references) {
          let reference = references[symbol];
          asset.addDependency({
            specifier: reference.specifier,
            specifierType: 'esm',
            symbols: new Map([[reference.name, {
              local: symbol,
              isWeak: false,
              loc: null
            }]])
          });
          asset.meta.hasReferences = true;
        }
      }

      assets.push({
        type: 'js',
        content: depjs + js,
        dependencies: [],
        env
      });
    }

    return assets;
  }

});

exports.default = _default;
let cache = new Map();

function getTargets(browsers) {
  if (browsers == null) {
    return undefined;
  }

  let cached = cache.get(browsers);

  if (cached != null) {
    return cached;
  }

  let targets = (0, _css().browserslistToTargets)((0, _browserslist().default)(browsers));
  cache.set(browsers, targets);
  return targets;
}
